Enforce state mutability in view pure checker.

ekpyron opened the pull request:
Originally part of #4090.

ekpyron committed:
Remove v050 check for enforcing "view" in ViewPureChecker.

ekpyron committed:
Adjust tests wrt strict view pure checker.

ekpyron committed:
Update documentation and Changelog.

ekpyron commented:
Surprisingly small amount of tests affected by this...

ekpyron commented:
Appveyor failure unrelated to this PR.

chriseth commented on docs/contracts.rst:
```diff
@@ -535,6 +532,9 @@ In addition to the list of state modifying statements explained above, the follo
   at runtime is actually of that type.
 
+.. warning::
+  Before version 0.5.0 the compiler didn't enforce that ``view`` is not writing the state.
```
Please use `did not`

chriseth commented on libsolidity/analysis/ViewPureChecker.cpp:
```diff
@@ -251,13 +242,8 @@ void ViewPureChecker::reportMutability(StateMutability _mutability, SourceLocati
-			m_errorReporter.typeError(_location, text);
-		}
+		m_errors = true;
+		m_errorReporter.typeError(_location, text);
```
Perhaps it would be better to move this up now (and delete the `text` variable)?

ekpyron replied:
True

chriseth commented on test/libsolidity/syntaxTests/viewPureChecker/write_storage_fail.sol:
```diff
@@ -3,4 +3,4 @@ contract C {
     function f() view public { x = 2; }
 }
 // ----
-// Warning: (56-57): Function declared as view, but this expression (potentially) modifies the state and thus requires non-payable (the default) or payable.
```
Shouldn't there be more tests where we have to change 'warning' for 'error' is this PR?

ekpyron replied:
It surprised me as well, but this seems to be the only one at the moment.

ekpyron replied:
Should I just add more tests in this PR?

chriseth replied:
Yes, please.

ekpyron committed:
Extend view pure test suite.

ekpyron commented:
@chriseth I added a number of new test cases. It seems that so far the tests mainly checked for the absence of the "can be restricted to ..." warnings. I renamed those test cases to make it clear what they are testing and for most of them added a copy with restricted mutability that expects failure.
Do you think we need even more test cases?

axic commented on docs/contracts.rst:
```diff
@@ -529,12 +526,15 @@ In addition to the list of state modifying statements explained above, the follo
   at runtime is actually of that type.
 
+.. warning::
+  Before version 0.5.0 the compiler did not enforce that ``view`` is not writing the state.
```
"writing to the state" or rather "modifying the state"?

axic commented on test/libsolidity/syntaxTests/viewPureChecker/selector.sol:
```diff
@@ -5,4 +5,8 @@ contract C {
         return this.f.selector ^ this.x.selector;
     }
+    function h() view public returns (bytes4) {
+        x;
```
Why is `x` needed here?
